### PR TITLE
Get rid of most of the RefCells in librustc_typeck/check

### PR DESCRIPTION
**Pull request by pythonesque**
There are two that I didn't touch that are kind of irksome to get rid of.  Though I do want to do it, I also don't want to make this commit too much larger.

I may also try to refactor this to see if I can get rid of some of the extra structures / function parameters; it was helpful as an intermediate step, but I think most of the time we can get away with just passing one context around.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Wow! Nice!

I don't personally work much in these crates though, so I'll defer to @nikomatsakis or @pnkfelix 

r? @pnkfelix

**Comment by arielb1** [comment]
couldn't you just take FnCtxt by mutable reference?

**Comment by pythonesque** [comment]
@arielb1 Not in general, because it has to be shared in a number of places (e.g. the `Typing` for the `euv` walker in `upvar.rs`).  With more refactoring it could be, though; I'm working on another patch to fix `trans` in a similar way to open up this possibility.

**Comment by arielb1** [comment]
I couldn't find any Typing in upvar.rs

**Comment by pythonesque** [comment]
@arielb1 https://github.com/rust-lang/rust/pull/25247/files#diff-193820a23727493def60049c0b6af47eR443 (might be called `Typer`).  The problem is that this needs to be able to access some shared variables (notably, the remaining `RefCell`'d ones) mutably, but the `Typer` implementation requires access as well.

**Comment by pythonesque** [comment]
You could still pass it as a mutable reference in most places, though, and just split it up into two internal structures when necessary (as it is there), which is the potential refactoring I alluded to above.  This is the approach I'm taking from the getgo in the trans patch.

**Comment by arielb1** [comment]
@pythonesque 

I am afraid your splitting creates lots of ugly duplication.
Well its called TYPER. I would add a method to get the `&mut TYPER` from the euv (store a TYPER rather than `&mut TYPER` and lift mc::Typer over `&` - I don't think euv relies on its typer staying the same.

**Comment by pythonesque** [comment]
@arielb1 Yes, like I said above, it can be done, but I didn't want to make the patch too much larger.

**Comment by arielb1** [comment]
@pythonesque 

The current patch seems rather large and ugly to me (passing FnCtxt and CheckEnv through all of typeck's methods - half of them already have rather more parameters than optimal).

**Comment by nikomatsakis** [comment]
@pythonesque 

I didn't have time to read this today, but I have been wanting to refactor a lot of these refcells away for some time, so this at least sounds great!

**Comment by pythonesque** [comment]
@arielb1 Just FYI, I just hit borrowck in my refactor on the trans side (that passes around a single &mut context) and got over 2100 errors, so I am not actually sure anymore whether that approach would be less ugly (the only way I see to deal with most of the errors is to immediately split out the parameters when they come into the function, and rejoin them every time they are passed in, which seems to me to largely defeat the purpose of trying to make it one parameter).  On the other hand, trans is probably a little different, since there we are dealing with multiple blocks, each of which references a single context.

**Comment by nikomatsakis** [comment]
I started a discuss thread to discuss this approach, since it seems to have some pros and cons:

https://internals.rust-lang.org/t/prs-removing-refcells-longer-term-plans/2099

@pythonesque I'm sorry for the slow feedback...it's been a bit crazy last few weeks.

File: src/librustc/middle/expr_use_visitor.rs [hunk 1]
```diff
@@ -1283,10 +1283,11 @@ impl<'d,'t,'tcx,TYPER:mc::Typer<'tcx>> ExprUseVisitor<'d,'t,'tcx,TYPER> {
...
 
-fn copy_or_move<'tcx>(typer: &mc::Typer<'tcx>,
-                      cmt: &mc::cmt<'tcx>,
-                      move_reason: MoveReason)
-                      -> ConsumeMode
+fn copy_or_move<'tcx,T>(typer: &T,
+                        cmt: &mc::cmt<'tcx>,
+                        move_reason: MoveReason)
+                        -> ConsumeMode
+    where T: mc::Typer<'tcx>
```
Comment (nikomatsakis) [inline src/librustc/middle/expr_use_visitor.rs]: Why make this more generic?

Comment (nikomatsakis) [inline src/librustc/middle/expr_use_visitor.rs]: Oh, I see, Typer is no longer object-safe. I think I'd prefer that Typer stay object-safe; pity we don't have a good solution for passing FnOnce() closurs as objects yet.

Comment (pythonesque) [inline src/librustc/middle/expr_use_visitor.rs]: Ah, yes, I remember the issue I had now.  Yeah, there was literally no way to add a closure (that I could see) without breaking object safety.

**Comment by pnkfelix** [comment]
@nikomatsakis @jroesch how would this change interact with the recently discussed revisions to the compiler infrastructure?  I am basically wondering if I should push through a review of this PR (and then presumably request it be rebased, assuming it passes review), or if we should put this aside for now until after other refactorings take place?

From what I remember, it seems like this PR should probably help more than it hurts future refactorings.

**Comment by nikomatsakis** [comment]
@pnkfelix While on vacation, I was thinking about these PRs and asking myself the same question. I'm not really clear on how these PR fit into the plans we were discussing. I think I have to get the various refactorings reloaded a bit into my head before I can evaluate the impact.

**Comment by jroesch** [comment]
I've been meaning to review this work as well, will sit down after lunch and read it over.

File: src/librustc/middle/expr_use_visitor.rs (hunk 1 above)
Comment (jroesch) [inline src/librustc/middle/expr_use_visitor.rs]: Doing a re-review of these changes, this doesn't matter anymore I axed the Typer interface last week.

**Comment by nikomatsakis** [comment]
Currently my feeling though is that we are pursuing a lot of refactorings -- flattening contexts, breaking up tables to be more local, incremental compilation, MIR -- and that it's we should leave things as they are under the assumption that they are going to be dismembered and put back together, and when we do this we will try to make best use of our type system.

**Comment by Gankro** [comment]
Based on the inactivity and @nikomatsakis' comment, I'm going to close this PR. 

Feel free to disagree!

**Pull request closed without merging**